File: client-sdks/csharp/sdk/VertexCacheSdk/Comm/ClientConnector.cs
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------------

using System;
using System.IO;
using System.Net.Sockets;
using System.Security.Cryptography;
using VertexCacheSdk.Model;

namespace VertexCacheSdk.Comm
{
    /// <summary>
    /// TcpClient is the core transport used by the VertexCache SDK to connect to and communicate with a VertexCache server.
    /// It manages socket creation, TLS negotiation, IDENT handshake, and encryption using AES-GCM or RSA.
    /// </summary>
    public class ClientConnector
    {
        private Stream _writer;
        private Stream _reader;
        private ClientOption _options;
        private bool _connected = false;
        private TcpClient _tcpClient;

        public ClientConnector(ClientOption options)
        {
            _options = options;
        }

        public void Connect()
        {
            try
            {
                if (_options.EnableTlsEncryption)
                {
                    var sslStream = SocketHelper.CreateSecureSocket(_options);
                    _writer = new BufferedStream(sslStream);
                    _reader = new BufferedStream(sslStream);
                }
                else
                {
                    var socket = SocketHelper.CreateSocketNonTLS(_options);
                    var stream = new NetworkStream(socket);
                    _writer = new BufferedStream(stream);
                    _reader = new BufferedStream(stream);
                }

                byte[] identPayload = EncryptIfEnabled(System.Text.Encoding.UTF8.GetBytes(_options.BuildIdentCommand()));
                MessageCodec.WriteFramedMessage(_writer, identPayload);
                _writer.Flush();

                byte[] identResponse = MessageCodec.ReadFramedMessage(_reader);
                string responseStr = identResponse == null ? "" : System.Text.Encoding.UTF8.GetString(identResponse).Trim();

                if (!responseStr.StartsWith("+OK"))
                {
                    throw new VertexCacheSdkException("Authorization failed: " + responseStr);
                }

                _connected = true;
            }
            catch (Exception ex)
            {
                throw new VertexCacheSdkException("Connection failed", ex);
            }
        }

        public string Send(string message)
        {
            lock (this)
            {
                try
                {
                    byte[] toSend = EncryptIfEnabled(System.Text.Encoding.UTF8.GetBytes(message));
                    MessageCodec.WriteFramedMessage(_writer, toSend);
                    _writer.Flush();

                    byte[] response = MessageCodec.ReadFramedMessage(_reader);
                    if (response == null)
                    {
                        throw new VertexCacheSdkException("Connection closed by server");
                    }

                    return System.Text.Encoding.UTF8.GetString(response);
                }
                catch (Exception ex)
                {
                    throw new VertexCacheSdkException("Unexpected failure during send", ex);
                }
            }
        }

        private byte[] EncryptIfEnabled(byte[] plainText)
        {
            try
            {
                switch (_options.EncryptionMode)
                {
                    case EncryptionMode.Asymmetric:
                        using (RSA rsa = KeyParserHelper.ConfigPublicKeyIfEnabled(_options.PublicKey))
                        {
                            return rsa.Encrypt(plainText, RSAEncryptionPadding.Pkcs1);
                        }

                    case EncryptionMode.Symmetric:
                        return GcmCryptoHelper.Encrypt(plainText, KeyParserHelper.ConfigSharedKeyIfEnabled(_options.SharedEncryptionKey));

                    case EncryptionMode.None:
                    default:
                        return plainText;
                }
            }
            catch (Exception ex)
            {
                throw new VertexCacheSdkException("Encryption failed for plaintext message", ex);
            }
        }

        public bool IsConnected()
        {
            return _connected;
        }

        public void Close()
        {
            try
            {
                _writer?.Close();
                _reader?.Close();
            }
            catch { }
            _connected = false;
        }
    }
}

File: client-sdks/kotlin/sdk/main/kotlin/com/vertexcache/sdk/comm/ClientConnector.kt
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------------

package com.vertexcache.sdk.comm

import com.vertexcache.sdk.model.EncryptionMode
import com.vertexcache.sdk.model.VertexCacheSdkException
import com.vertexcache.sdk.model.ClientOption
import java.io.*
import java.net.Socket
import java.security.PublicKey
import javax.crypto.Cipher

/**
 * ClientConnector is the core transport used by the VertexCache SDK to connect to and communicate with a VertexCache server.
 * It manages socket creation, TLS negotiation (with optional certificate verification), client authentication via IDENT,
 * and encryption of messages using either symmetric (AES-GCM) or asymmetric (RSA) methods.
 */
class ClientConnector(private val options: ClientOption) {

    private var socket: Socket? = null
    private var writer: BufferedOutputStream? = null
    private var reader: BufferedInputStream? = null
    private var connected: Boolean = false

    /**
     * Establishes a connection and performs IDENT handshake.
     * Uses TLS or plain socket depending on configuration.
     */
    fun connect() {
        try {
            socket = if (options.enableTlsEncryption) {
                SocketHelper.createSocketTLS(options)
            } else {
                SocketHelper.createSocketNonTLS(options)
            }

            writer = BufferedOutputStream(socket!!.getOutputStream())
            reader = BufferedInputStream(socket!!.getInputStream())

            val identBytes = encryptIfEnabled(options.buildIdentCommand().toByteArray())
            MessageCodec.writeFramedMessage(writer!!, identBytes)
            writer!!.flush()

            val response = MessageCodec.readFramedMessage(reader!!)
            val identResponse = response?.toString(Charsets.UTF_8)?.trim() ?: ""
            if (!identResponse.startsWith("+OK")) {
                throw VertexCacheSdkException("Authorization failed: $identResponse")
            }
            connected = true
        } catch (e: Exception) {
            throw VertexCacheSdkException("Connection failed", e)
        }
    }

    /**
     * Sends a message and waits for framed response.
     */
    @Synchronized
    fun send(message: String): String {
        try {
            val payload = encryptIfEnabled(message.toByteArray())
            MessageCodec.writeFramedMessage(writer!!, payload)
            writer!!.flush()
            val response = MessageCodec.readFramedMessage(reader!!)
                ?: throw VertexCacheSdkException("Connection closed by server")
            return String(response, Charsets.UTF_8)
        } catch (e: Exception) {
            throw VertexCacheSdkException("Unexpected failure during send", e)
        }
    }

    /**
     * Applies encryption based on configured mode.
     */
    private fun encryptIfEnabled(plainText: ByteArray): ByteArray {
        return try {
            when (options.encryptionMode) {
                EncryptionMode.ASYMMETRIC -> {
                    MessageCodec.switchToAsymmetric()
                    val cipher = Cipher.getInstance("RSA")
                    cipher.init(Cipher.ENCRYPT_MODE, options.getPublicKeyAsObject())
                    cipher.doFinal(plainText)
                }
                EncryptionMode.SYMMETRIC -> {
                    MessageCodec.switchToSymmetric()
                    GcmCryptoHelper.encrypt(plainText, options.getSharedEncryptionKeyAsBytes())
                }
                EncryptionMode.NONE -> plainText
            }
        } catch (e: Exception) {
            throw VertexCacheSdkException("Encryption failed for, text redacted *****", e)
        }
    }

    /**
     * Checks connection state.
     */
    fun isConnected(): Boolean {
        return connected && socket != null && socket!!.isConnected &&
                !socket!!.isClosed && !socket!!.isInputShutdown && !socket!!.isOutputShutdown
    }

    /**
     * Closes socket.
     */
    fun close() {
        try {
            socket?.close()
        } catch (_: IOException) {
        }
        connected = false
    }
}

